### ISSUE TITLE
Output parameters vs return values

### ISSUE THREAD

**Issue by dimztimz** [eliaskosunen/scnlib#129]
My comments are mostly for the C++ proposal P1729. In the [first revision](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1729r1.html) the API used output parameters, similar to `scanf` and `iostreams`, but in the [second revision](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p1729r2.html) it was changed to use return value. Now, the newer API might be more ergonomic (arguable), but it can be less performant, especially when reading strings.

Consider the following example:

```c++
std::string key;
int value;
for (...) {
    std::scan(some_input, "{} = {}", key, value);
}
```

versus

```c++
for (...) {
    auto result = std::scan<std::string, int>(some_input, "{} = {}")
}
```

The first example is more performant because it will reuse the same `std::string` object and it will avoid unnecesarry allocations. In the second example the `std::string` object will be constructed and destructed in each loop iteration, potentially allocating in each iteration. SSO saves something, but in the first example we save much more unnecessary allocations.

**Comment by TD22057** [comment]
I agree w/ this comment.  IMO the v1 API was simpler to use, easier to understand, and easier to see how the format string tied to values.  Converting our code to the new version has resulting in much longer and more confusing parse statements.

**Comment by dimztimz** [comment]
Ping, any comments about this? One should take into account the user feedback too, not only the LEWGI feedback. The API design of v1 is better using output parameters for the data and [return value for the error](https://github.com/eliaskosunen/scnlib/tree/v1?tab=readme-ov-file#error-handling). The only nitpicking can be done about the API for the error class.

Alternatively, the proposal can propose both APIs, the with output params and another with return vale.

**Comment by LapNik** [comment]
I would definitely prefer v2. Output parameters always get clunky in my experience. C++ Core Guidelines authors seem to agree: [F.20](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-out)

**Comment by dimztimz** [comment]
> I would definitely prefer v2. Output parameters always get clunky in my experience. C++ Core Guidelines authors seem to agree: [F.20](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-out)

At the end of the guideline you can see few exceptions to it. The following is relevant:

> Exceptions:
> - ...
> - To reuse an object that carries capacity (e.g., std::string, std::vector) across multiple calls to the function in an inner loop: [treat it as an in/out parameter and pass by reference](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-out-multi).
